src/fcc_bridge_node_safety.cpp

```diff
@@ -795,79 +751,89 @@ bool FCCBridgeNode::check_point_in_geofence(const double latitude_deg,
+                    static_cast<double>(speed_mps),
+                    static_cast<double>(this->safety_limits->max_speed_mps));
+        return false;
+    }
+
     return true;
 }
 
 void FCCBridgeNode::check_last_mission_control_heartbeat() {
     RCLCPP_DEBUG(this->get_safety_logger(),
                  "Checking last mission control heartbeat");
 
-    switch (this->get_internal_state()) {
-            // This function should never be called in these states
-        case INTERNAL_STATE::ERROR:
-            // This should never happen, as the process exits on ERROR state
-            throw std::runtime_error(std::string(__func__) +
-                                     " called while in ERROR state");
-        case INTERNAL_STATE::STARTING_UP:
-        case INTERNAL_STATE::ROS_SET_UP:
+    if (this->get_internal_state() == INTERNAL_STATE::STARTING_UP ||
+        this->get_internal_state() == INTERNAL_STATE::ROS_SET_UP) {
+        RCLCPP_FATAL(this->get_internal_state_logger(),
+                     "In an invalid state for a mission control heartbeat "
+                     "check check! Exiting...");
+        this->set_internal_state(INTERNAL_STATE::ERROR);
+        this->exit_process_on_error();
+    }
+
+    if (!this->last_mission_control_heartbeat.has_value()) {
+        if (this->get_internal_state() == INTERNAL_STATE::MAVSDK_SET_UP) {
+            RCLCPP_WARN(this->get_safety_logger(),
+                        "Got no cached mission control heartbeat. "
+                        "Acceptable while in state %s",
+                        this->internal_state_to_str());
+            return;
+        } else if (this->is_airborne()) {
+            RCLCPP_ERROR(
+                this->get_safety_logger(),
+                "Got no cached mission control heartbeat! Triggering RTH...");
+            this->trigger_rth();
+            return;
+        } else {
+            RCLCPP_FATAL(this->get_safety_logger(),
+                         "Got no cached mission control heartbeat! EXITING...");
+            this->set_internal_state(INTERNAL_STATE::ERROR);
+            this->exit_process_on_error();
+        }
+    }
+
+    if (rclcpp::Duration(MAX_MISSION_CONTROL_HEARTBEAT_AGE) <
+        (this->now() - this->last_mission_control_heartbeat->time_stamp)) {
+        if (this->get_internal_state() == INTERNAL_STATE::RETURN_TO_HOME ||
+            this->get_internal_state() == INTERNAL_STATE::LANDED) {
+            RCLCPP_WARN(
+                this->get_safety_logger(),
+                "Mission control timed out while in state %s. Ignoring...",
+                this->internal_state_to_str());
+            return;
+        }
+        if (this->is_airborne()) {
             RCLCPP_FATAL(
-                this->get_internal_state_logger(),
-                "In an invalid state for a uav health check! Exiting...");
+                this->get_safety_logger(),
+                "The last mission control heartbeat is older than: %" PRId64
```

Comment:
Aren't we missing a comma after PRId64 here?

Reply:
No, PRId64 is a macro defined in the libc header [cinttypes](https://en.cppreference.com/w/c/types/integer) which expands to the format specifier for a signed 64 bit integer. On my particular machine it is defined as "d". The compiler will then merge the adjacent string literals into one.
Because fixed width integers use different types depending on a platform (e.g. a int32_t might be a int or a long etc.) and the format specifiers are based on these traditional types (int = %d, long = %ld), just specifying %d is not portable.
Thats why I am using these Macros. Otherwise the compiler might issue a warning on the PI

Reply:
I knew why you are using it, I was just wondering if you missed a comma afterwards. But if it works, then great!